src/wp-includes/post.php

```diff
@@ -70,7 +70,7 @@ function create_initial_post_types() {
 	);
 
 	// Enhance page editor for block themes by rendering template and content blocks.
-	if ( wp_is_block_theme() && current_theme_supports( 'block-templates' ) ) {
+	if ( did_action( 'after_setup_theme' ) && wp_is_block_theme() && current_theme_supports( 'block-templates' ) ) {
```

Comment:
[We noticed](https://github.com/WordPress/wordpress-develop/pull/8123#issuecomment-2677186673) when reviewing the original PR that if `wp_is_block_theme()` is called too soon, it would cause side effects that caused tests to fail.

Looking at the code places in Core that call `create_initial_post_types()`, it looks like this get called in `wp-settings.php` before the themes are set up, which can have the same affect.

Adding a check to ensure `after_setup_theme` has run seems to fix this issue. However, there's also potential for third-party code to call `wp_is_block_theme()` too early, so we may want to add that check to `wp_is_block_theme()` itself and throw a `_doing_it_wrong`.

Reply:
If the intent is that calling `wp_is_block_theme` shouldn't happen until after theme setup, than I absolutely agree a `_doing_it_wrong` makes sense. I also think this should be a reasonable assumption.